cirq/value/product_state.py

```diff
@@ -0,0 +1,319 @@
+        """
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            oneq_proj = oneq_state.projector()
+            mat = np.kron(mat, oneq_proj)
+        return mat
+
+
+class _PauliEigenState(_NamedOneQubitState):
+
+    def __init__(self, eigenvalue):
+        self.eigenvalue = eigenvalue
+        self._eigen_index = (1 - eigenvalue) / 2
+
+    @property
+    @abc.abstractmethod
+    def _symbol(self):
+        pass
+
+    def __str__(self):
+        sign = {1: '+', -1: '-'}[self.eigenvalue]
+        return f'{sign}{self._symbol}'
+
+    def __repr__(self):
+        return f'cirq.{self._symbol}.basis[{self.eigenvalue:+d}]'
+
+    @abc.abstractmethod
+    def stabilized_by(self):
+        pass
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        return self.eigenvalue == other.eigenvalue
+
+    def __hash__(self):
+        return hash((self.__class__.__name__, self.eigenvalue))
+
+    def _json_dict_(self):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return protocols.obj_to_dict_helper(self, ['eigenvalue'])
+
+
+class _XEigenState(_PauliEigenState):
+    _symbol = 'X'
+
+    def state_vector(self):
+        if self.eigenvalue == 1:
+            return np.array([1, 1]) / np.sqrt(2)
+        elif self.eigenvalue == -1:
+            return np.array([1, -1]) / np.sqrt(2)
+        # coverage: ignore
+        raise ValueError("Bad eigenvalue: {}".format(self.eigenvalue))
+
+    def stabilized_by(self):
+        # Prevent circular import from `value.value_equality`
+        from cirq import ops
+        return self.eigenvalue, ops.X
+
+
+class _YEigenState(_PauliEigenState):
+    _symbol = 'Y'
+
+    def state_vector(self):
+        if self.eigenvalue == 1:
+            return np.array([1, 1j]) / np.sqrt(2)
+        elif self.eigenvalue == -1:
+            return np.array([1, -1j]) / np.sqrt(2)
+        # coverage: ignore
+        raise ValueError("Bad eigenvalue: {}".format(self.eigenvalue))
+
+    def stabilized_by(self):
+        from cirq import ops
+        return self.eigenvalue, ops.Y
+
+
+class _ZEigenState(_PauliEigenState):
+    _symbol = 'Z'
+
+    def state_vector(self):
+        if self.eigenvalue == 1:
+            return np.array([1, 0])
+        elif self.eigenvalue == -1:
+            return np.array([0, 1])
+        # coverage: ignore
+        raise ValueError("Bad eigenvalue: {}".format(self.eigenvalue))
+
+    def stabilized_by(self):
+        from cirq import ops
+        return self.eigenvalue, ops.Z
+
+
+KET_PLUS = _XEigenState(eigenvalue=+1)
```

Comment:
These symbols lead to two different ways of accessing the 1-qubit stabilizer states: by names (when available) and by the ".basis" trick. This goes against the [Zen of python](https://www.python.org/dev/peps/pep-0020/). If adherence to the Zen isn't a sufficient argument against them, then consider that they will also lead to expressions that mix the two ways - these will look ugly - and may delay the users becoming familiar with the ".basis" trick - which we want them to learn as soon as possible.

Suggestion: Remove all KET_* symbols (but keep `PAULI_STATES` - this is useful, e.g. for iterating over all 1-qubit stabilizer states).

Reply:
(Relevant excerpt from the Zen: "There should be one-- and preferably only one --obvious way to do it.")

Reply:
I see the current construct analogous to how we have both `cirq.CNotPowGate(exponent=1)` and `cirq.CNOT`. I also think there's educational value in presenting e.g. |+> as either the 0'th basis element of the X pauli operator or "the plus state"